fix(lsp): ignore null responses for semanticTokens request

fsouza opened the pull request:
The spec indicates that the response may be `null`, but it doesn't really say what a `null` response means. Since neovim raises an error if the response is `null`, I figured that ignoring it would be the safest bet.

fsouza commented:
I didn't open an issue because this seems simple enough, but I see the following error when using the kotlin-language-server with nvim `v0.9.0-dev-486+g3cf0131c5`:

```
Error executing vim.schedule lua callback: ...eovim/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:310: attempt to index local 'response' (a nil value)
stack traceback:
        ...eovim/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:310: in function 'process_response'
        ...eovim/share/nvim/runtime/lua/vim/lsp/semantic_tokens.lua:269: in function 'handler'
        ...isco_souza/opt/neovim/share/nvim/runtime/lua/vim/lsp.lua:1389: in function ''
        vim/_editor.lua: in function <vim/_editor.lua:0>
```

mfussenegger reviewed (commented):
Could you also add a test case?

jdrouhard commented:
Must have missed that in the spec. Thanks for catching and the quick fix @fsouza

Can you put the check below the reset of active_request, though? A nil response is still a response.

theHamsta commented:
I think I got sometimes `NULL` responses from old versions sumneko_lua when it was not fully initialized

fsouza committed:
fix(lsp): ignore null responses for semanticTokens request

The spec indicates that the response may be `null`, but it doesn't
really say what a `null` response means. Since neovim raises an error if
the response is `null`, I figured that ignoring it would be the safest
bet.

fsouza committed:
[squash later] add a test for #21364

fsouza commented:
Just pushed a test. I pushed it as a separate commit to make it easier to review, I'll squash it if folks think it looks good. I had to modify _create_server to make sure handlers can provide a `nil` response.

The easiest way to do that without potentially breaking too many tests was to pass the callback to the handler. Let me know if that's not a good idea.

mfussenegger committed:
fixup! [squash later] add a test for #21364

mfussenegger commented:
> The easiest way to do that without potentially breaking too many tests was to pass the callback to the handler. Let me know if that's not a good idea.

I changed it to just always use the response, even if it's null to avoid having two different ways to send results back
